File: 509.fibonacci-number.java
/*
 * @lc app=leetcode id=509 lang=java
 *
 * [509] Fibonacci Number
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    public int fib(int n) {
        List<Integer> table = new ArrayList<>();
        table.add(0);
        table.add(1);
        int idx = 2;
        while (idx <= n) {
            table.add(table.get(idx - 1) + table.get(idx - 2));
            idx += 1;
        }
        return table.get(n);
    }
}
// @lc code=end

File: 509.fibonacci-number.cpp
/*
 * @lc app=leetcode id=509 lang=cpp
 *
 * [509] Fibonacci Number
 */

// @lc code=start

#include <vector>

using namespace std;
class Solution {
public:
    int fib(int n) {
        vector<int> fibs(max(n + 1, 2), 0);
        fibs[0] = 0; fibs[1] = 1;
        for (int ix = 2; ix <= n; ++ix) {
            fibs[ix] = fibs[ix - 1] + fibs[ix - 2];
        }
        return fibs[n];
    }
};
// @lc code=end
